src/Containers/Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -339,68 +343,186 @@ public async Task<string> DownloadBlobAsync(string repository, Descriptor descri
         return localPath;
     }
 
-    public async Task PushAsync(Layer layer, string repository, Action<string> logProgressMessage, CancellationToken cancellationToken)
+    public async Task PushAsync(Layer layer, string repository, Action<LogMessage> logProgressMessage, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
         string digest = layer.Descriptor.Digest;
 
         using (FileStream contents = File.OpenRead(layer.BackingFile))
         {
-            await UploadBlobAsync(repository, digest, contents, cancellationToken).ConfigureAwait(false);
+            await UploadBlobAsync(repository, digest, contents, logProgressMessage, cancellationToken).ConfigureAwait(false);
         }
     }
 
-    private async Task<UriBuilder> UploadBlobChunkedAsync(string repository, string digest, Stream contents, HttpClient client, UriBuilder uploadUri, CancellationToken cancellationToken)
+    /// <summary>
+    /// Captures the data needed to finalize an upload
+    /// </summary>
+    private record UploadFinalizeInformation(UriBuilder uploadUri, string digest);
+
+    /// <summary>
+    /// Upload a blob to the registry using chunks. The chunks are uploaded sequentially, and the size of the chunks is determined by the UploadInformation derived from the 'Start' call
+    /// </summary>
+    private async Task<UploadFinalizeInformation> UploadBlobChunkedAsync(string repository, Stream contents, HttpClient client, UploadInformation uploadInfo, Action<LogMessage> logProgressMessage, CancellationToken cancellationToken)
```

Comment:
This method changed a lot, but I refactored it to hopefully be more clear.  It's a state machine, so I've modeled it as a series of states based on the status code of the response we get back from uploading each chunk.

The counts for retry logic are very much negotiable - I started with amounts from golang clients that I researched for background here.

Essentially we start a processing loop:
* while there are chunks to send
  * read a chunk
  * send the chunk
  * if ok, loop (there's a quirk here for AWS)
  * if error < retryMax, reset our read state and try again, incrementing our counter
  * finally error out

each of these decisions are a separate method so they can be inspected as individuals.